Creating shipment line item records when shipment is delivered

reggieriser opened the pull request:
## Description

This PR covers two stories linked below, both which create `ShipmentLineItem` records when a shipment is delivered.  The first story creates a single line item for the `LHS` (linehaul) code.  The second story creates three line items for `135A` (origin service fee), `135B` (destination service fee), and `105A` (full pack -- although we include unpack here for now -- see note below).

## Reviewer Notes

We do not currently have a `tariff400ng_items` code for the unpack line item, so I have combined the pack/unpack into a single line item with the costs added together (I believe this is the way it used to be done).  I am making a separate story for dealing with that issue.

Also note that PR #1338 needs to be merged to master prior to this PR to avoid some UI confusion.  More details below.

## Setup

To test in the app, you will need to get a shipment to the "in transit" state, making sure the shipment has a net weight and an actual pickup date.  You should be able to populate your DB with the `db_populate_e2e` script, then do a local login to the TSP interface with `tspuser1@example.com`.  Go to the "in transit" shipments in the queue, then pick the shipment with locator `ENTDEL`.  Click the `Deliver` button and submit a delivery date.  You should now see four new records in the `ShipmentLineItem` table.  They will also appear for the moment in the "Pre-Approval Requests" panel, but that's a bug that's being fixed in #1338.  That PR should be merged prior to this PR.

To be safe, I like to use the UI to get a shipment into the "in transit" state, then backup my database and restore it as needed for testing (this takes more time, but feels like a little safer test than using E2E generated data).

The server tests will also validate these `ShipmentLineItem` entries on shipment delivery.

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](./docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](./docs/backend.md#querying-the-database-safely)
 have been satisfied.
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story 1](https://www.pivotaltracker.com/story/show/161693995) for this change
* [Pivotal story 2](https://www.pivotaltracker.com/story/show/161564001) for this change
* [Pivotal story](https://www.pivotaltracker.com/story/show/161944267) for missing `Full Unpack` code

kahlouie commented on pkg/models/shipment.go:
```diff
@@ -703,3 +701,68 @@ func SaveShipmentAndAddresses(db *pop.Connection, shipment *Shipment) (*validate
+			verrs, err = s.createUniqueShipmentLineItem(tx, lineItem)
+			if err != nil || verrs.HasAny() {
+				responseVErrors.Append(verrs)
+				responseError = errors.Wrap(err, "Error saving shipment line item")
```
Should we include some data about which shipment line item errored, or is that already part of `err`?

reggieriser replied:
The most likely value of `err` will have some context, but it's probably worth adding some here just in case.  Just committed some code to do that.

reggieriser commented on pkg/models/shipment.go:
```diff
@@ -503,13 +503,11 @@ func FetchShipmentByTSP(tx *pop.Connection, tspID uuid.UUID, shipmentID uuid.UUI
 		"Move.Orders.NewDutyStation.Address",
-		"Move.Orders.HasDependents",
-		"Move.Orders.SpouseHasProGear",
-		"Move.Orders.ServiceMemberID",
```
You're welcome!  And for anybody else who's wondering why these lines were removed ... you only eager fetch relationships, not individual columns.  So eager fetching `Move.Orders` would suffice to bring in all direct fields of `Orders`.  However, since we're already eagerly fetching even further down the hierarchy with `Move.Orders.NewDutyStation.Address`, that covers `Orders` for us.

stangah commented on pkg/models/shipment.go:
```diff
@@ -703,3 +701,75 @@ func SaveShipmentAndAddresses(db *pop.Connection, shipment *Shipment) (*validate
+	err := db.Q().
+		Where("shipment_id = ?", s.ID).
+		Where("tariff400ng_item_id = ?", tariff400ngItemID).
+		All(&lineItems)
```
Instead of `All`, there's a `First` method that I think would simplify things a bit here if you want

reggieriser replied:
My first inclination was to try `First` (I've used it before), but I believe it throws an error if the record isn't there.  I didn't want to try to distinguish between that error and other types of errors that may be tossed, so I went with an `All` which won't throw errors if no matching records exist.  I tried using an `Exists` too at some point in the process and ran into issues, but I may try again as I'm not seeing why that wouldn't work now given how things turned out.

stangah replied:
Ah makes sense. I totally get not wanting to fish out that error, for a long time we used string comparison to do it and it felt bad. But for future reference if you wanted you can compare against the actual underlying error which you can grab from the `database/sql` library, `ErrNoRows` https://golang.org/pkg/database/sql/#pkg-variables

You also need to run it through `errors.Cause` to _actually_ get the root error in that case, but it's there :p

akostibas commented on pkg/handlers/publicapi/shipments_test.go:
```diff
@@ -730,8 +732,131 @@ func (suite *HandlerSuite) TestDeliverShipmentHandler() {
+		RateRefCode:         models.Tariff400ngItemRateRefCodeTARIFFSECTION,
+		RequiresPreApproval: false,
+	}
+	suite.MustSave(&codeLHS)
```
It might be a small improvement to move these item approvals to the test data gen package:
https://github.com/transcom/mymove/blob/master/pkg/testdatagen/make_tariff400ng_items.go

Something to consider, not necessarily needed.

akostibas commented on pkg/models/shipment.go:
```diff
@@ -703,3 +701,68 @@ func SaveShipmentAndAddresses(db *pop.Connection, shipment *Shipment) (*validate
+		return nil, nil
+	}
+
+	return &lineItems[0], nil
```
Actually, it looks like this is a duplicate: https://github.com/transcom/mymove/blob/master/pkg/models/shipment_line_item.go#L81

reggieriser replied:
Agree that returning the whole slice may be better for future use.  Will adjust.

I don't think the other function is doing the same thing.  It looks to be finding a shipment line item by its ID (independent of shipment ID).  In my case, I had a different use case -- I needed a way to find out if a specific shipment had a line item for a specific item ID so I knew whether to allow another to be created (I didn't know/care what the ID for the existing shipment line item itself was).

akostibas commented on pkg/models/shipment_test.go:
```diff
@@ -217,3 +218,60 @@ func (suite *ModelSuite) TestCreateShipmentLineItem() {
+			Tariff400ngItem:   item,
+		}
+		lineItems = append(lineItems, lineItem)
+	}
```
Could you use this approach to making test data above in `pkg/handlers/publicapi/shipments_test.go`? So much more concise!

reggieriser replied:
Maybe so.  In the other case, I believe I was trying to be as precise with all the fields as I could because it felt like they mattered more there for testing -- so doing something similar to the above didn't have the same conciseness.  But some of the values in the other case may be irrelevant to testing correctness.  I'll see.

akostibas commented on pkg/handlers/publicapi/shipments.go:
```diff
@@ -325,17 +327,109 @@ func (h DeliverShipmentHandler) Handle(params shipmentop.DeliverShipmentParams)
+		return handlers.ResponseForError(h.Logger(), err)
+	}
+
+	var lineItems []models.ShipmentLineItem
```
I was having the same thought. I think my vote would be in the RateEngine, mostly to prevent circular dependencies (e.g. handler -> model -> rateengine -> model)

reggieriser replied:
Agreed.  I had previously tried to move this into the models package but got bit by the circular dependency @akostibas mentioned.  I'll see if I can refactor it into the rateengine package instead.

reggieriser commented on pkg/models/shipment.go:
```diff
@@ -703,3 +701,75 @@ func SaveShipmentAndAddresses(db *pop.Connection, shipment *Shipment) (*validate
+
+	if existingLineItem != nil {
+		var whichCode string
+		if len(lineItem.Tariff400ngItem.Code) > 0 {
```
It's just a protection against that method being called in the future with a ShipmentLineItem that hasn't eagerly fetched the Tariff400ngItem association.  We'd have a "zero" struct in that case.  I would prefer to show the Code in the error, but didn't feel like I wanted to mandate the method require the association be loaded just for the error.

reggieriser commented:
Made a few changes based on feedback:

* Moved creation (and testing) of base ShipmentLineItem records to the rateengine package.
* Moved creation of tariff test data (and a few other fields) into the testdatagen package so that any callers of `CreateShipmentOfferData` get that related data on an in-transit or delivered shipment.
* Refactored `FetchShipmentLineItemByItemID` to `FetchShipmentLineItemsByItemID` so that it returned all line items of a particular type for a shipment.
